### PR TITLE
For ApproximateSizes, pro-rate table metadata size over data blocks

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: The implementation of GetApproximateSizes was inconsistent in
its treatment of the size of non-data blocks of SST files, sometimes
including and sometimes now. This was at its worst with large portion
of table file used by filters and querying a small range that crossed
a table boundary: the size estimate would include large filter size.

It's conceivable that someone might want only to know the size in terms
of data blocks, but I believe that's unlikely enough to ignore for now.
Similarly, there's no evidence the internal function AppoximateOffsetOf
is used for anything other than a one-sided ApproximateSize, so I intend
to refactor to remove redundancy in a follow-up commit.

So to fix this, GetApproximateSizes (and implementation details
ApproximateSize and ApproximateOffsetOf) now consistently include in
their returned sizes a portion of table file metadata (incl filters
and indexes) based on the size portion of the data blocks in range. In
other words, if a key range covers data blocks that are X% by size of all
the table's data blocks, returned approximate size is X% of the total
file size. It would technically be more accurate to attribute metadata
based on number of keys, but that's not computationally efficient with
data available and rarely a meaningful difference.

Also includes miscellaneous comment improvements / clarifications.

Also included is a new approximatesizerandom benchmark for db_bench.
No significant performance difference seen with this change, whether ~700 ops/sec with cache_index_and_filter_blocks and small cache or ~150k ops/sec without cache_index_and_filter_blocks.

Test Plan: Test added to DBTest.ApproximateSizesFilesWithErrorMargin.
Old code running new test...

    [ RUN      ] DBTest.ApproximateSizesFilesWithErrorMargin
    db/db_test.cc:1562: Failure
    Expected: (size) <= (11 * 100), actual: 9478 vs 1100

Other tests updated to reflect consistent accounting of metadata.

### CODE REVIEW

File: table/block_based/block_based_table_reader.cc [hunk 1]
```diff
@@ -2979,26 +2979,41 @@ Status BlockBasedTable::CreateIndexReader(
...
+    assert(!rep_->filter_handle.IsNull());
+    return rep_->filter_handle.offset();
+  }
+  // Index handle not readily available, so...
+  if (rep_->compression_dict_handle != BlockHandle()) {
+    assert(!rep_->compression_dict_handle.IsNull());
+    return rep_->compression_dict_handle.offset();
+  }
+  // Hmm, use rough estimate from footer
+  return rep_->footer.metaindex_handle().offset();
```
Comment (siying) [inline table/block_based/block_based_table_reader.cc:3010]: rep_->table_properties is only null if the file was generated by release some early 2.x releases or earlier. This is a rare case. My vote would be to simplify the case. We don't have to try filter and directly go to metaindex. And I believe dictionary compression was not available for those early versions.

File: table/block_based/block_based_table_reader.cc [hunk 2]
```diff
@@ -3037,13 +3060,19 @@ uint64_t BlockBasedTable::ApproximateSize(const Slice& start, const Slice& end,
...
+  uint64_t start_offset = ApproximateDataOffsetOf(*index_iter, data_size);
   index_iter->Seek(end);
-  uint64_t end_offset = ApproximateOffsetOf(*index_iter);
+  uint64_t end_offset = ApproximateDataOffsetOf(*index_iter, data_size);
 
   assert(end_offset >= start_offset);
-  return end_offset - start_offset;
+  // Pro-rate file metadata (incl filters) size-proportionally across data
+  // blocks.
+  double size_ratio = static_cast<double>(end_offset - start_offset) / static_cast<double>(data_size);
```
Comment (siying) [inline table/block_based/block_based_table_reader.cc:3074]: This line is over 80-char.

File: table/block_based/block_based_table_reader.cc [hunk 3]
```diff
@@ -2979,26 +2979,41 @@ Status BlockBasedTable::CreateIndexReader(
...
+  }
+  // Hmm, use rough estimate from footer
+  return rep_->footer.metaindex_handle().offset();
+}
 
-  return result;
+uint64_t BlockBasedTable::GetApproximateFileSize() {
+  // This is close enough without keeping around the exact size
+  BlockHandle mih = rep_->footer.metaindex_handle();
+  return mih.offset() + mih.size() + Footer::kMaxEncodedLength;
```
Comment (siying) [inline table/block_based/block_based_table_reader.cc:3016]: I know it gets more work but can we remember file_size by passing it from BlockBasedTable::Open()?
The footer structure might be applicable to change. For example, we might introduce padding before super block and footer for various reasons (e.g. encryption), and this hack makes it slightly prone to bugs.

Comment (pdillinger) [reply table/block_based/block_based_table_reader.cc:3016]: I'm in favor of that if another 8 bytes per open file is considered negligible

Comment (siying) [reply table/block_based/block_based_table_reader.cc:3016]: I think 8 bytes are fine. It's already a huge object.

**Pull request closed without merging**